### PR TITLE
Alteration: Knights Fighting.

### PR DESCRIPTION
**Pull request by lobodInI**

### CODE REVIEW

File: app/great_battle.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from app.structuring_the_knight_personality import Knight
+
+
+class BattleOfTwoKnight:
```
Comment (Dimosphen1) [inline app/great_battle.py:4]: Pick up the class name without prepositions, the class name should be as laconic as possible.

File: app/great_battle.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    @staticmethod
+    def battle_action_phase(knight_1: Knight, knight_2: Knight) -> None:
+
+        damage_for_first_knight = knight_2.power - knight_1.protection
+        damage_for_second_knight = knight_1.power - knight_2.protection
+        knight_1.hp -= max(0, damage_for_first_knight)
+        knight_2.hp -= max(0, damage_for_second_knight)
+
+    @staticmethod
+    def result_after_battle(knight: Knight) -> None:
```
Comment (Dimosphen1) [inline app/great_battle.py:14]: Method names should describe an action, do not use single-standing nouns or phrases as method names. You should utilize verbs here and in another method.

File: app/preparing_for_the_great_battle.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+from app.structuring_the_knight_personality import Knight
+
+
+class BeforeBattle:
```
Comment (Dimosphen1) [inline app/preparing_for_the_great_battle.py:4]: The class name `BeforeBattle` does not make much sense; in case you use it only as a function collection or a collection of static methods, think of writing separate functions and collecting them in separate `support_functons.py` file.

File: app/structuring_the_knight_personality.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from typing import List, Dict, Optional
+
+
+class Knight:
```
Comment (Dimosphen1) [inline app/structuring_the_knight_personality.py:4]: Think of adding methods here that could be beneficial to this class in order not to use it as a data structure only.

File: app/structuring_the_knight_personality.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+from typing import List, Dict, Optional
```
Comment (Dimosphen1) [inline app/structuring_the_knight_personality.py:1]: Rename the file, it is too wordy to store one `Knight` class.

File: app/great_battle.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+from app.structuring_the_knight_personality import Knight
+
+
+class BattleOfTwoKnight:
+    @staticmethod
```
Comment (Dimosphen1) [inline app/great_battle.py:5]: In case you use it only as a function collection or a collection of static methods, think of writing separate functions and collecting them in separate files.

**Review by Dimosphen1 — CHANGES_REQUESTED** [review]
Several changes were requested. Split the project into different modules and packages as well.

File: app/main.py [hunk 7]
```diff
@@ -1,214 +1,37 @@
...
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
+from app.legendary_knights import KNIGHTS
+from app.structuring_the_knight_personality import Knight
+from app.preparing_for_the_great_battle import BeforeBattle
+from app.great_battle import BattleOfTwoKnight
 
-    # -------------------------------------------------------------------------------
-    # BATTLE:
+# ______________________WELCOME TO THE TOURNAMENT__________________________
```
[the remark this answers is not in the document]
Comment (lobodInI) [reply app/main.py:8]: Ok.

File: app/main.py [hunk 8]
```diff
@@ -1,214 +1,37 @@
...
-        lancelot["hp"] = 0
+def battle(knights: Dict[str, Any] = KNIGHTS) -> Dict[str, int]:
 
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
+    lancelot = Knight(**knights["lancelot"])
+    arthur = Knight(**knights["arthur"])
+    mordred = Knight(**knights["mordred"])
+    red_knight = Knight(**knights["red_knight"])
+    # list of registered knights for the duel
```
[the remark this answers is not in the document]
Comment (lobodInI) [reply app/main.py:17]: 👍

File: app/main.py [hunk 9]
```diff
@@ -1,214 +1,37 @@
...
+    list_knight = [lancelot, arthur, mordred, red_knight]
```
[the remark this answers is not in the document]
Comment (lobodInI) [reply app/main.py:18]: 👍

File: app/legendary_knights.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
+KNIGHTS = {
```
[the remark this answers is not in the document]
Comment (lobodInI) [reply app/legendary_knights.py:1]: 👍

File: app/great_battle.py (hunk 1 above)
Comment (lobodInI) [reply app/great_battle.py:4]: 👍
just_be_creative

File: app/great_battle.py (hunk 2 above)
Comment (lobodInI) [reply app/great_battle.py:14]: 👍

File: app/preparing_for_the_great_battle.py (hunk 3 above)
Comment (lobodInI) [reply app/preparing_for_the_great_battle.py:4]: 👍

File: app/structuring_the_knight_personality.py (hunk 4 above)
Comment (lobodInI) [reply app/structuring_the_knight_personality.py:4]: 👍
just_be_creative

File: app/structuring_the_knight_personality.py (hunk 5 above)
Comment (lobodInI) [reply app/structuring_the_knight_personality.py:1]: 👍
just_be_creative

File: app/structuring_the_knight_personality.py (hunk 4 above)
Comment (lobodInI) [reply app/structuring_the_knight_personality.py:4]: merge preparation_battle.py and knight_class.py?

then what is the meaning of the packages?

File: app/great_battle.py (hunk 6 above)
Comment (lobodInI) [reply app/great_battle.py:5]: 👍

File: app/Knights/knight_class.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
+from typing import List, Dict, Optional
```
Comment (Dimosphen1) [inline app/Knights/knight_class.py:1]: Rename your module to use lower-letter naming in a single form.

File: app/Battle/preparation_battle.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
+from app.Knights.knight_class import Knight
```
Comment (Dimosphen1) [inline app/Battle/preparation_battle.py:1]: Rename your module to use lower-case form.

File: app/Battle/great_battle.py [hunk 13]
```diff
@@ -0,0 +1,14 @@
+from app.Knights.knight_class import Knight
```
Comment (Dimosphen1) [inline app/Battle/great_battle.py:1]: There is no need to use the word `great` in file naming, it does not contribute to a better understanding of its content. You could use `support_functions.py` or `battle_functions.py` here.

File: app/Battle/great_battle.py [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+
+def battle_action_phase(knight_1: Knight, knight_2: Knight) -> None:
```
Comment (Dimosphen1) [inline app/Battle/great_battle.py:4]: You should use verbs for function naming to describe an action that is performed by a function. Rename it to be more descriptive. You could use `perform_fight` here as an example.

File: app/Battle/preparation_battle.py [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
```
Comment (Dimosphen1) [inline app/Battle/preparation_battle.py:2]: Rename your module to `preparation_functions.py` or another name to be more descriptive of its content.

File: app/main.py [hunk 16]
```diff
@@ -1,214 +1,35 @@
...
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
+    lancelot = Knight(**knights["lancelot"])
+    arthur = Knight(**knights["arthur"])
+    mordred = Knight(**knights["mordred"])
+    red_knight = Knight(**knights["red_knight"])
 
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
+    knights_names = [lancelot, arthur, mordred, red_knight]
```
Comment (Dimosphen1) [inline app/main.py:16]: This `knights_names` list contains not names but rather `Knight` instances, rename it to `knights`.

**Review by Dimosphen1 — CHANGES_REQUESTED** [review]
Several changes were requested.

File: app/Knights/knight_class.py (hunk 11 above)
Comment (lobodInI) [reply app/Knights/knight_class.py:1]: 👍

File: app/Battle/preparation_battle.py (hunk 12 above)
Comment (lobodInI) [reply app/Battle/preparation_battle.py:1]: 👍

File: app/Battle/great_battle.py (hunk 13 above)
Comment (lobodInI) [reply app/Battle/great_battle.py:1]: 👍

File: app/Battle/great_battle.py (hunk 14 above)
Comment (lobodInI) [reply app/Battle/great_battle.py:4]: 👍

File: app/Battle/preparation_battle.py (hunk 15 above)
Comment (lobodInI) [reply app/Battle/preparation_battle.py:2]: 👍

File: app/main.py (hunk 16 above)
Comment (lobodInI) [reply app/main.py:16]: 👍

File: app/main.py [hunk 17]
```diff
@@ -1,214 +1,35 @@
...
-        arthur["name"]: arthur["hp"],
-        mordred["name"]: mordred["hp"],
-        red_knight["name"]: red_knight["hp"],
+        knight.name: knight.hp
+        for knight in knights
     }
 
 
-print(battle(KNIGHTS))
+print(battle())
```
Comment (Bohdan-Salamakha) [inline app/main.py:35]: remove this test print function

**Review by Bohdan-Salamakha — CHANGES_REQUESTED** [review]

**Review by Bohdan-Salamakha — APPROVED** [review]
